### DISCUSSION TITLE
Two-switch scanning on dynamic displays

### DISCUSSION THREAD

**Discussion by gbishop** [UNC-Project-Open-AAC/OS-DPI#96] [Q&A]
How do systems typically handle two-switch scanning on dynamic displays? Ours is potentially very dynamic changing the size of a grid, or switching tabs in a tab control. 

When the layout changes do systems start scanning from the top?

I could imagine scenarios where coming back to a tab starts you where you left off but starting at the top seems more likely. 

Or consider the case of the [grid with morphs](https://unc-project-open-aac.github.io/OS-DPI/?fetch=examples/updated/grid_ex_2.osdpi). I've used my switches to move over to "go". I select it once and then again and get the morphs. Scanning will step through go, goes, went, gone, going. After I select one it goes back to the home screen. Where should be scan be now? It could be on "go" where you left off. It could start at the upper left. 

Consider the keyboard in that same app. Maybe it starts with choosing between the completion grid and the keyboard grid. I choose the keyboard and move over to the "d". Does it stay on the d? Or immediately go back to the top and offer me the choice between completions and letters? How do I get to the completions if it stays on the d?

I'm guessing designers are going to want the ability to do any of these...

What do current systems allow and what is reasonable for us to allow?

**Comment by cdsjeff** [comment]
back to the top for scanning is probably the most common approach. Vanderheiden suggested that it was overall more efficient than proceding forward.  Its predictable for the user and offers some advantages like quick access to speak, backspace and delete functions.
J

**Reply by gbishop** [reply]
I think I would hate a keyboard that started over at the top after each selection especially if I needed to type a double letter. Typing Mississippi would be hell.

**Reply by lageist** [reply]
Yes. This could be a case where the option to stay within the group with a frequency organized keyboard arrangement would be good to consider.

**Reply by sofia-bg** [reply]
Here are some of the features built into iOS for scanning. I had a client that liked being able to toggle through these different patterns - especially the point scanning instead of having to wait for cycling through column row scanning. https://support.apple.com/en-us/HT201370

**Comment by lageist** [comment]
Back to the top is the default on systems. Many also include the option to enable a Scan from Last setting. Other options to consider: skip empty buttons, set custom patterns for groups (e.g., 2 rows of prediction buttons, 3 rows of letters treated as 2 groups), number of passes before restarting scan pattern.

**Comment by cdsjeff** [comment]
Right!

**Comment by gbishop** [comment]
> set custom patterns for groups (e.g., 2 rows of prediction buttons, 3 rows of letters treated as 2 groups),

You'll need to expand on these.

**Reply by lageist** [reply]
Scan patterns offered include: row/column, column/row, and linear. Most systems also offer the option to define groups. Groups might share attributes like part of speech (e.g., pronouns, adjectives, verbs) or function (letters, prediction buttons, speech controls, navigation to topic pages). The groupings tend to be obvious where organizing by group for scanning makes sense to consider. You define the order in which you want the groups to be scanned and once a group is selected, your scan pattern is applied within the group. I can dig up some specific UI examples if that would be helpful.

**Reply by gbishop** [reply]
> I can dig up some specific UI examples if that would be helpful.
Yes please.

**Reply by lageist** [reply]
Word Power by Nancy Inman and published on several commercial products includes use of groups (aka regions, blocks) in some of the scanning configurations (e.g., [Tobii-Dynavox example](https://tobii-dynavox.secure.force.com/publicweb/servlet/rtaImage?eid=ka13W000000TOwU&feoid=00Nw0000006l3S0&refid=0EMw0000000J4G5), [Saltillo example](https://saltillo.com/images/wordpower140-scan-touch.jpg)). [CoreScanner from PRC](https://www.prentrom.com/assets/uploads/corescanner-r2k.jpg) also uses groups.

**Comment by sofia-bg** [comment]
It seems with the morph example that could be a research question for someone? So you could try both approaches, I am not sure how you are envisioning the rules for applying scanning would be - but you could have to approaches one where regardless of what was selected in the morph you returned to top OR one where "if the last word selected was a morph then resume from there" I am not sure you would necessarily want to do that, seems just as likely your next word precedes go. 

Words with double letters is an interesting conundrum. I think that might also be a limitation of standard QWERTY keyboard and might be resolved with a either a frequency keyboard or a keyboard that not only has a prediction row but also mutes all buttons that could not possible be next and skips them when resuming scanning? Just to suggest speed might be improved by combination of access and layout? Wasn't this "problem" part of the reason for the advent of T9?

**Comment by cdsjeff** [comment]
I don’t disagree, but you’re talking about a statistical exception here.
Statistically speaking, it’s probably more efficient to start at the top. I
think you can solve your Mississippi problem with word prediction, making
that one of the first options for a scanner to select. We’re prediction is
a great choice for someone that asked rely on scanning.
J

**Reply by sofia-bg** [reply]
I used to have a car that had a clunky navigation system with a turn dial, but to speed it up as soon as you selected a letter it greyed out the letters that could not possibly come next and when you spun it didn't even select the greyed out letters just jumped to the next possible. Letters were arranged alphabetically. Seems like it could be an interesting option to test - maybe on an abc not qwerty keyboard?

**Comment by gbishop** [comment]
We can simulate that now I think.
